core/sail/lucene-api/src/main/java/org/eclipse/rdf4j/sail/lucene/LuceneSail.java

```diff
@@ -296,6 +296,13 @@ public class LuceneSail extends NotifyingSailWrapper {
 	 */
 	public static final String MAX_DOCUMENTS_KEY = "maxDocuments";
 
+	/**
+	 * Set the key "maxQueryDocuments=&lt;n&gt;" as sail parameter to limit the maximum number of documents the user can
+	 * query at a time to return from a search query. The default is the value of the {@link #MAX_DOCUMENTS_KEY}
+	 * parameter.
+	 */
+	public static final String MAX_QUERY_DOCUMENTS_KEY = "maxQueryDocuments";
```

Comment:
You have two parameters:
1. `MAX_DOCUMENTS_KEY`
2. `MAX_QUERY_DOCUMENTS_KEY`

The first one is the original one, it defines the maximum amount of documents the index should search. But with the new predicate added in this PR, maybe the user would set a limit above this value.

This limit can be both a feature (We increase temporary the amount of results for a query) or a security issue (Because we are removing a limit). So I decided to add the 2nd parameter, this one defines the maximum amount of documents the user can ask using the new predicate.

By default, I took the same value as `MAX_DOCUMENTS_KEY` because maybe someone will update RDF4J and isn't aware of this change, so he won't set the new parameter.

Reply:
Can you instead have only MAX_DOCUMENTS_KEY and then you can provide a lower default from the client side if the user doesn't provide a value?

So instead of something like:

```
max docs = 1000
max query docs = 5000
num docs = -1
```

You can instead do 

```
max docs = 5000
num docs = 1000 (client side sets num docs to 1000 as default and user can increase it to 5000 if needed, or set a lower value)
```

Reply:
That would require you to add the logic to your own application.

Reply:
Benefit would be that the MAX_DOCUMENTS setting would always be the maximum, even when the user specifies numDocs.

Reply:
Actually, I would want this PR to fix the issue https://github.com/ec-doris/kohesio-backend/issues/268, which requires overwriting the original max docs parameter at query time, both above and below its original value.

Also, it might break previous configs in other projects used to the previous MAX_DOCUMENTS logic, so I'm not sure about this way.

Reply:
The previous MAX_COUNT logic wouldn't change. It would still be the absolute upper limit. But if you want to allow users to go above this limit, then you would have to increase it and then specify your own numDocs on your requests if you still want some queries to return fewer than MAX_COUNT by default.

Reply:
How about introducing DEFAULT_NUM_DOCS that would be used instead of MAX_COUNT, unless `numDocs` is provided by the user.

So:
 - DEFAULT_NUM_DOCS = Math.min(MAX_COUNT, System.getProperty(...))
 - if numDocs is not provided by the user, use DEFAULT_NUM_DOCS
 - else limit numDocs to MAX_COUNT (e.g. Math.min(numDocs, MAX_COUNT)

Reply:
The benefit is that we only have a single MAX deal with, instead of two different MAXs depending on if the user provides a numDocs or not.

Reply:
So if I understand correctly, it would ask to replace all the queries to increase the count for only one of them?

Reply:
No. You would set DEFAULT_NUM_DOCS as a system property, instead of MAX_QUERY_DOCUMENTS.

Reply:
I am not sure to understand, because it sounds like a renaming of what I propose with the value of `MAX_QUERY_DOCUMENTS` also applied to `MAX_DOCUMENTS`.

`MAX_DOCUMENTS` becomes `DEFAULT_NUM_DOCS` and `MAX_QUERY_DOCUMENTS` `MAX_DOCUMENTS`.

Reply:
Yes. That's essentially it. That way we only have one maximum, but it still allows you to increase the maximum without affecting other queries by setting the default. 

My reason is that I found it confusing to have two configs that were both maximum number of documents, but applied I slight different scenarios. And I think it might be confusing for others in the future.

Reply:
But if we rename the `MAX_DOCUMENTS` param, won't it break some codes by changing the API?

An idea would be to rename only MAX_QUERY_DOCUMENTS, but I'm not sure what would be the best for you.

Reply:
No. We keep MAX_DOCUMENTS, and if a user provides their own numDocs at query time then it can't be higher than MAX_DOCUMENTS. 

If MAX_DOCUMENTS=9 and the user provides numDocs=3, then we return up to 3 documents. 

If you want to allow users to return more documents then you need to increase MAX_DOCUMENTS. 

Example: MAX_DOCUMENTS=12. Now a user can provide numDocs=12 and get 12 results back. 

But if the user doesn't provide a numDocs then we would still be returning (up to) 12 documents. 

This is where I want to introduce DEFAULT_NUM_DOCS. If the user doesn't provide a numDocs we return up to DEFAULT_NUM_DOCS. To be backwards compatible we set DEFAULT_NUM_DOCS=MAX_DOCUMENTS, unless the user explicitly configures their own value for DEFAULT_NUM_DOCS.

Now we can set DEFAULT_NUM_DOCS=9 in the config and set MAX_DOCUMENTS=12. If the user provides a numDocs of 50, we still only return (up to) 12 results, if the user provides numDocs=1 then we only return (up to) 1 result. If the user doesn't provide a numDocs we return up to 9 results, because DEFAULT_NUM_DOCS is 9.

Reply:
Ok as you want, my concern what about replacing all the MAX_DOCUMENTS references to DEFAULT_NUM_DOCS to use the new feature. But if you think it is better.